orttraining/orttraining/core/framework/gradient_graph_builder.cc

```diff
@@ -127,16 +131,49 @@ Status GradientGraphBuilder::Build() {
   ORT_RETURN_IF_ERROR(opt_ret);
 
   GraphAugmenter::GraphDefs gradient_graph_defs;
-  // add "gradient of the loss" node, always 1.
-  if (loss_node_arg_name_ != "") {
+
+  // We skip adding loss gradient in case it's partitioned by pipeline.
+  if(COMMUNICATION_NODES.find(graph_->GetProducerNode(loss_node_arg_name_)->OpType()) == COMMUNICATION_NODES.end()) {
```

Comment:
This seems conflicting with pipeline design. We add loss gradient even if the loss is output of `Send`.

Reply:
Ok that's fine. But regarding the loss gradient, it's always been an initializer of value 1, I'm not sure the value of adding this initializer at every boundary of partition.

Reply:
Great. Let's split this PR. We can discussion how to create a clean interface between my code and your code for the gradient builder part. 

For non-last pipeline stage, our code uses output of Send as loss value. If you mean `why do we add loss value as an initializer?`, then I don't know. Loss value should be graph output, not an initializer.

Reply:
updated.